src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Variables.cs

```diff
@@ -0,0 +1,373 @@
+                    }
+                }
+            }
+
+            private Variables GetVariablesForVariable(VariableIdentifier identifier)
+            {
+                int containingSlot = identifier.ContainingSlot;
+                if (containingSlot > 0)
+                {
+                    return GetVariablesForId(DeconstructSlot(containingSlot).Id)!;
+                }
+                return GetVariablesContainingVariable(identifier.Symbol);
+            }
+
+            internal Variables GetVariablesContainingVariable(Symbol symbol)
+            {
+                switch (symbol)
+                {
+                    case LocalSymbol:
+                    case ParameterSymbol:
+                    case MethodSymbol { MethodKind: MethodKind.LocalFunction }:
+                        if (symbol.ContainingSymbol is MethodSymbol method &&
+                            GetVariablesForSymbol(method.PartialImplementationPart ?? method) is { } variables)
+                        {
+                            return variables;
+                        }
+                        break;
+                }
+                return GetVariablesRoot();
+            }
+
+            private Variables GetVariablesRoot()
+            {
+                var variables = this;
+                while (variables.Container is { } container)
+                {
+                    variables = container;
+                }
+                return variables;
+            }
+
+            private Variables? GetVariablesForId(int id)
+            {
+                var variables = this;
+                do
+                {
+                    if (variables.Id == id)
+                    {
+                        return variables;
+                    }
+                    variables = variables.Container;
+                }
+                while (variables is { });
+                return null;
+            }
+
+            private Variables? GetVariablesForSymbol(MethodSymbol symbol)
+            {
+                var variables = this;
+                while (true)
+                {
+                    if (variables.Symbol is null || (object)symbol == variables.Symbol)
+                    {
+                        return variables;
+                    }
+                    variables = variables.Container!;
+                    if (variables is null)
+                    {
+                        return null;
+                    }
+                }
+            }
+
+            private int GetSlotDepth(int slot)
+            {
+                int depth = 0;
+                while (slot > 0)
+                {
+                    depth++;
+                    slot = this[slot].ContainingSlot;
+                }
+                return depth;
+            }
+
+            internal static int ConstructSlot(int id, int index)
+            {
+                return index < 0 ? index : (id << IdOffset) + index;
```

Comment:
>+ [](start = 60, length = 1)

Consider using ```|``` instead of ```+``` since we are operating with bits. It might also be good to assert that bits are not going to overlap.